### ISSUE TITLE
Ceramic Schema reference

### ISSUE THREAD

**Issue by PaulLeCam** [ceramicnetwork/CIP#82]
```
cip: to be assigned
title: Ceramic Schema reference
author: Paul Le Cam (@PaulLeCam)
status: Draft
category: Standards
type: RFC
created: 2021-02-15
```
  

## Simple Summary

Provide a static way to define that Ceramic documents referenced in a schema must use a specified schema or set of schemas.

## Abstract

This CIP defines a standard way to add references to existing JSON schemas, so it is possible to access this information about Ceramic documents at build time rather than only at runtime on created documents.

## Motivation

IDX definitions contain a `schema` property that references an existing schema, guarantying the IDX record associated to the definition matches this schema.
This CIP provides a way to implement similar logic for any schema independently of IDX.

Using this CIP, a `NotesList` schema could explicitly reference a `Note` schema with the following example:

```
{
  $schema: 'http://json-schema.org/draft-07/schema#',
  title: 'Notes',
  type: 'object',
  properties: {
    notes: {
      type: 'array',
      title: 'list',
      items: {
        type: 'object',
        title: 'item',
        properties: {
          note: {
            type: 'object',
            $id: 'ceramic://docSchema',
            title: 'reference',
            properties: {
              schema: { type: 'string', const: '<Note schema docID>' },
              id: { type: 'string', maxLength: 150 },
            },
          },
          title: {
            type: 'string',
            maxLength: 100,
          },
        },
        required: ['note'],
      },
    },
  },
}
```
This way, by loading the `Notes` schema, it is possible by a tool/library to discover the `Note` schema the same way loading a IDX definition allows the discovery of the record's schema.

## Specification

References to Ceramic schema should use an `object` with the [`$id`](https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.2.2) field of `ceramic://docSchema`, and with a `schema` property containing either a [`const`](https://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.6.1.3) (single schema) or non-empty [`enum`](https://json-schema.org/draft/2019-09/json-schema-validation.html#enum) (multiple schemas) field referencing the target schema(s), as defined below:

```js
{
  type: 'object',
  $id: 'ceramic://docSchema',
  title: 'mySchemaDoc',
  properties: {
    schema: { type: 'string', const: '<schema docID>' },
    id: { type: 'string', maxLength: 150 },
  },
}
```

The `title` field is optional and can be filled with any relevant value, and additional properties can be added to the object.

## Rationale

This CIP builds on [CIP81](https://github.com/ceramicnetwork/CIP/issues/81), providing additional metadata (the schema) for static discovery.

This spec allows to either define a single schema (using `const`) or multiple ones (using `enum`).
The use case would be to support different schemas for a single reference, for example a "media" schema could reference an "image" schema, but also the "audio" and "video" ones as acceptable document schemas: `{ type: 'string', enum: ['<image schema docID>', '<audio schema docID>', '<video schema docID>'] }`.

## Backwards Compatibility

As we don't have schema references other than IDX definitions this CIP should not require any change to our existing schemas.

## Implementation

None yet.

## Security Considerations

None I'm aware of.

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by oed** [comment]
This seems pretty great!

A few questions:

Maybe naming the CIP **DocID json-schema definition** or something similar that inlcudes *json-schmea* would be more clear?

Why would we want to include multiple schemas in one reference? Would it not be better to have multiple definitions for each schema to decrease ambiguity?
 
I think the example would be more clear if you used a named definition!

**Comment by PaulLeCam** [comment]
> This seems pretty great!

Thanks!

> Maybe naming the CIP **DocID json-schema definition** or something similar that inlcudes _json-schmea_ would be more clear?

Yes good idea 👍 

> Why would we want to include multiple schemas in one reference? Would it not be better to have multiple definitions for each schema to decrease ambiguity?

At build time it's used to identify possible schemas the document could use, but it's not necessarily restricted to a single one.
For example we can have a  media feed document that has items that can be of different types (image, audio or video) so at built time we can discover that we need to be aware of the image, audio and video schemas, but at runtime the loaded document will specify its schema anyways so we can have the runtime logic work accordingly.
The main use case here is similar to the [GraphQL union type](https://spec.graphql.org/June2018/#sec-Unions).

> I think the example would be more clear if you used a named definition!

Sorry I'm not sure what you mean by named definition here?

**Comment by oed** [comment]
> 

Like so:
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "title": "IDXKeychain",
  "properties": {
    "notes": {
      "type": "array",
      "items": { "$ref": "#/definitions/DocID" }
    }
  },
  "additionalProperties": false,
  "required": [ "notes" ],
  "definitions": {
    "DocID": {
      "type": "string",
      "$id": "ceramic://doc",
      "$ceramicSchema": "<schema docID>",
      "maxLength": 150,
    }
  }
}
```

I think this makes it a bit clearer for someone who is trying to understand what this is :)

**Comment by PaulLeCam** [comment]
I don't think it makes much sense to use the `definitions` here because using `$ceramicSchema` makes the field specific to a given schema rather than generic to any document, so in this case this would be simpler:

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "notes": {
      "type": "array",
      "items": {
        "type": "string",
        "$id": "ceramic://doc",
        "$ceramicSchema": "<schema docID>",
        "maxLength": 150
      }
    }
  },
  "additionalProperties": false,
  "required": ["notes"]
}
```

**Comment by oed** [comment]
Yeah, that's simpler. I more meant that it would be easier to understand for someone reading this CIP because it clearly separates what is part of the DocID reference.

**Comment by PaulLeCam** [comment]
I see, thanks! I updated the example.

**Comment by oed** [comment]
`'<Note schema docID>'` should be `'<Note schema commitID>'` right?

**Comment by PaulLeCam** [comment]
I should accept either, maybe we can name it `docRef` then as per Sergey's PR?

**Comment by oed** [comment]
Schemas always have to be a CommitID afaik.

**Comment by oed** [comment]
@PaulLeCam Any blockers for making a PR for this CIP?

**Comment by PaulLeCam** [comment]
No blocker, I opened a PR for it that references this issue for discussions: https://github.com/ceramicnetwork/CIP/pull/86

**Comment by oed** [comment]
I'm seeing an issue if I have multiple docids in the same schema:
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "title": "ConvictionState",
  "properties": {
    "context": {
      "type": "string"
    },
    "supply": {
      "type": "number"
    },
    "participants": {
      "type": "array",
      "items": {
        "$ref": "#/definitions/participants"
      }
    },
    "proposals": {
      "type": "array",
      "items": {
        "$ref": "#/definitions/proposals"
      }
    }
  },
  "additionalProperties": false,
  "required": [
    "context",
    "supply",
    "participants",
    "proposals"
  ],
  "definitions": {
    "participants": {
      "type": "object",
      "properties": {
        "account": {
          "type": "string"
        },
        "balance": {
          "type": "number"
        },
        "convictions": {
          "type": "string",
          "$id": "ceramic://doc",
          "$ceramicSchema": "<Convictions commitid>",
          "maxLength": 150
        }
      },
      "required": [
        "account",
        "balance",
        "convictions"
      ]
    },
    "proposals": {
      "type": "object",
      "properties": {
        "proposal": {
          "type": "string",
          "$id": "ceramic://doc",
          "$ceramicSchema": "<Proposal docid>",
          "maxLength": 150
        },
        "totalConviction": {
          "type": "number"
        },
        "triggered": {
          "type": "boolean"
        }
      },
      "required": [
        "proposal",
        "totalConviction",
        "triggered"
      ]
    }
  }
}
```

At https://www.jsonschemavalidator.net/ I see this error:

<img width="451" alt="image" src="https://user-images.githubusercontent.com/3909429/109419570-6afb2580-79ce-11eb-9319-fb27d5ac2325.png">

**Comment by PaulLeCam** [comment]
Seems like we can't use `$id` as I intended then, thanks for reporting this, I'll look for alternative options.

**Comment by PaulLeCam** [comment]
@oed I created https://github.com/ceramicnetwork/CIP/issues/88 and opened a PR to update CIP-82 with these changes in https://github.com/ceramicnetwork/CIP/pull/89, I think that would nicely work around this issue.
